powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/totp/TOTP.java

```diff
@@ -0,0 +1,299 @@
+    private static boolean validateTOTP(final String otp, final byte[] key, final LocalDateTime localDateTime, final int steps, final String algorithm) throws CryptoProviderException {
+        logger.debug("Validating TOTP for localDateTime={}, algorithm={}, steps={}", localDateTime, algorithm, steps);
+
+        if (otp == null || otp.trim().length() == 0) {
+            throw new CryptoProviderException("Otp is mandatory");
+        }
+
+        if (steps < 0) {
+            throw new CryptoProviderException("Steps must not be negative number");
+        }
+
+        long currentTimeStep = countTimeSteps(localDateTime);
+        for (int i = 0; i <= steps; i++) {
+            logger.debug("Validating TOTP for localDateTime={}, algorithm={}, backward step={} out of allowed steps={}", localDateTime, algorithm, i, steps);
+            final long step = currentTimeStep - i;
+            final String expectedOtp = generateTOTP(key, step, otp.length(), algorithm);
+            if (expectedOtp.equals(otp)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Generates a TOTP value for the given set of parameters.
+     *
+     * @param key          the shared secret
+     * @param timeStep     number of time step
+     * @param returnDigits number of digits to return
+     * @param algorithm    the algorithm to use
+     * @return a numeric String in base 10 that includes truncation digits
+     * @throws CryptoProviderException in case of any crypto error
+     */
+    private static String generateTOTP(final byte[] key, final long timeStep, final int returnDigits, final String algorithm) throws CryptoProviderException {
+        logger.debug("Generating TOTP for timeStep={}, algorithm={}", timeStep, algorithm);
+
+        if (key == null) {
+            throw new CryptoProviderException("Key is mandatory");
+        }
+
+        if (algorithm == null) {
+            throw new CryptoProviderException("Algorithm is mandatory");
+        }
+
+        if (returnDigits <= 0) {
+            throw new CryptoProviderException("ReturnDigits must be positive number");
+        }
+
+        // Using the counter
+        // First 8 bytes are for the movingFactor
+        // Compliant with base RFC4226 (HOTP)
+        final String timeWithCounterPrefix = padWithZeros(Long.toHexString(timeStep), 16);
+
+        // Get the HEX in a Byte[]
+        final byte[] msg = HexFormat.of().parseHex(timeWithCounterPrefix);
+
+        final byte[] hash = computeHash(algorithm, key, msg);
+
+        // put selected bytes into result int
+        final int offset = hash[hash.length - 1] & 0xf;
+
+        final int binary = ((hash[offset] & 0x7f) << 24) |
+                ((hash[offset + 1] & 0xff) << 16) |
+                ((hash[offset + 2] & 0xff) << 8) |
+                (hash[offset + 3] & 0xff);
+
+        final int otp = binary % DIGITS_POWER[returnDigits];
+
+        final String result = Integer.toString(otp);
```

Comment:
Is there a reason for using String here? I would prefer using `byte[]` for such operations and only convert the result to String, if required.